src/event/finalization.rs

```diff
@@ -0,0 +1,336 @@
+#![deny(missing_docs)]
+
+use atomig::{Atom, AtomInteger, Atomic, Ordering};
+use serde::{Deserialize, Serialize};
+use std::{mem, sync::Arc};
+use tokio::sync::oneshot;
+
+type ImmutVec<T> = Box<[T]>;
+
+/// Wrapper type for an array of event finalizers. This is the primary
+/// public interface to event finalization metadata.
+#[derive(Clone, Debug, Default)]
+pub struct EventFinalizers(ImmutVec<Arc<EventFinalizer>>);
+
+impl Eq for EventFinalizers {}
+
+impl PartialEq for EventFinalizers {
+    fn eq(&self, other: &Self) -> bool {
+        self.0.len() == other.0.len()
+            && (self.0.iter())
+                .zip(other.0.iter())
+                .all(|(a, b)| Arc::ptr_eq(a, b))
+    }
+}
+
+impl EventFinalizers {
+    /// Create a new array of event finalizer with the single event.
+    pub fn new(finalizer: EventFinalizer) -> Self {
+        Self(vec![Arc::new(finalizer)].into())
+    }
+
+    /// Merge the given list of finalizers into this array.
+    pub fn merge(&mut self, other: Self) {
+        if !other.0.is_empty() {
+            // This requires a bit of extra work both to avoid cloning
+            // the actual elements and because `self.0` cannot be
+            // mutated in place.
+            let finalizers = mem::replace(&mut self.0, vec![].into());
+            let mut result: Vec<_> = finalizers.into();
+            // This is the only step that may cause a (re)allocation.
+            result.reserve_exact(other.0.len());
+            // Box<[T]> is missing IntoIterator
+            let other: Vec<_> = other.0.into();
+            for entry in other.into_iter() {
```

Comment:
```suggestion
            for entry in (&other.0).into_iter() {
```

Should do without copying over the vec.

Reply:
I was pretty careful here to avoid any copies, and the suggestion doesn't work. Calling `into_iter` on a reference generates an iterator over _references_ to the elements, requiring a `clone` to store the entries as values. `Box<[T]>` into `Vec<T>` is an in-place transform, just adding a `capacity` element.